typing_tests/pine-options.ts

```diff
@@ -0,0 +1,624 @@
+> = {
+	$expand: [
+		{
+			owns__device: {
+				$filter: {
+					device_name: null,
+					note: 'note',
+					device_environment_variable: {
+						$any: {
+							$alias: 'dev',
+							$expr: {
+								dev: {
+									name: 'name',
+									// this should be failing, but it isn't :(
+									// See: https://github.com/microsoft/TypeScript/issues/32000
+									asdf: 'asdf', // $ExpectError-Skip
+								},
+							},
+						},
+					},
+				},
+				$expand: {
+					device_environment_variable: {
+						$filter: {
+							name: 'name',
+							value: null,
+							asdf: 'asdf', // $ExpectError
+						},
+					},
+				},
+			},
+		},
+		'depends_on__application',
+	],
+};
+
+// valid selects
+
+export const appOptionsEValid1: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$select: ['app_name'],
+};
+
+export const deviceOptionsEValid2: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Device
+> = {
+	$select: ['note', 'device_name', 'uuid'],
+};
+
+// valid expands
+
+export const appOptionsEValid2: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$expand: {
+		owns__device: {
+			$select: ['note', 'device_name', 'uuid'],
+		},
+	},
+};
+
+export const appOptionsEValid4: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$expand: {
+		owns__device: {
+			$expand: {
+				device_environment_variable: {
+					$select: ['name', 'value'],
+				},
+			},
+		},
+	},
+};
+
+export const appOptionsEValid5: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$expand: {
+		owns__device: {
+			$select: ['note', 'device_name', 'uuid'],
+			$expand: {
+				device_environment_variable: {
+					$select: ['name', 'value'],
+				},
+			},
+		},
+	},
+};
+
+export const appOptionsEValid20: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$expand: [
+		{
+			owns__device: {
+				$select: ['note', 'device_name', 'uuid'],
+				$expand: {
+					device_environment_variable: {
+						$select: ['name', 'value'],
+					},
+				},
+			},
+		},
+		'depends_on__application',
+	],
+};
+
+// valid filters
+
+export const appOptionsEValidf1: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$filter: {
+		app_name: 'test',
+		slug: null,
+		application_type: 5,
+	},
+};
+
+export const appOptionsEValidf2: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$filter: {
+		app_name: 'test',
+		slug: null,
+		application_type: {
+			$any: {
+				$alias: 'o',
+				$expr: {
+					o: {
+						name: 'test',
+					},
+				},
+			},
+		},
+		owns__device: {
+			$any: {
+				$alias: 'd',
+				$expr: {
+					d: {
+						device_name: 'test',
+					},
+				},
+			},
+		},
+		owns__release: {
+			$any: {
+				$alias: 'r',
+				$expr: {
+					1: 1,
+				},
+			},
+		},
+		$not: {
+			owns__release: {
+				$any: {
+					$alias: 'r',
+					$expr: {
+						r: { commit: 'commit' },
+					},
+				},
+			},
+		},
+	},
+};
+
+export const appOptionsEFValid1: BalenaSdk.PineOptionsFor<
+	BalenaSdk.Application
+> = {
+	$expand: [
```

Comment:
Would it work if we have a single (or 2-3) complete query with all possible combinations inside it, and we just add multiple error assertions there? That should dramatically reduce the amount of code, and still achieve the same result.

Reply:
No unfortunately not :(
TS either
* reports the error on the outermost parent (you can see the `nestedArray$expandWithManyErrors` tests)
* or stops further reporting errors for the expression as soon as an error is found, which makes the rest expectations to error with `Expected an error on this line, but found none.`

Reply:
Except if I'm missing/forgetting something, since this needed considerable effort, in which case PRs are of course welcome :)